File: frontend/src/app/core/services/strategy.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Position {
  symbol?: string;
  segment: string;
  expiry: string;
  strike?: number;
  option_type?: string;
  action: string;
  qty: number;
  entry_price: number;
  iv?: number;
  delta?: number;
  theta?: number;
  gamma?: number;
  vega?: number;
}

export interface Strategy {
  id?: number;
  name: string;
  symbol: string;
  positions: Position[];
  created_at?: string;
}

export interface UserStrategies {
  live: Strategy[];
  history: Strategy[];
}

export interface PortfolioSyncedHolding {
  id: number;
  symbol: string;
  instrument_type: 'EQUITY' | 'FUTURE' | 'OPTION';
  qty: number;
  lots?: number;
  lot_size?: number;
  avg_price: number;
  invested: number;
  expiry?: string | null;
  strike?: number | null;
  option_type?: 'CE' | 'PE' | null;
  action?: 'BUY' | 'SELL' | null;
  notes?: string | null;
  created_at?: string;
  updated_at?: string;
}

export interface PortfolioSyncedResponse {
  count: number;
  total_invested: number;
  holdings: PortfolioSyncedHolding[];
}

export interface YearwiseDataPoint {
  yesterday_chng_per?: number;
  one_week_chng_per?: number;
  one_month_chng_per?: number;
  three_month_chng_per?: number;
  six_month_chng_per?: number;
  one_year_chng_per?: number;
  two_year_chng_per?: number;
  three_year_chng_per?: number;
  five_year_chng_per?: number;
  index_yesterday_chng_per?: number;
  index_one_week_chng_per?: number;
  index_one_month_chng_per?: number;
  index_three_month_chng_per?: number;
  index_six_month_chng_per?: number;
  index_one_year_chng_per?: number;
  index_two_year_chng_per?: number;
  index_three_year_chng_per?: number;
  index_five_year_chng_per?: number;
  index_name?: string;
}

export interface YearwiseDataResponse {
  symbol: string;
  identifier: string;
  count: number;
  data: YearwiseDataPoint[];
}

@Injectable({
  providedIn: 'root'
})
export class StrategyService {
  private apiUrl = `${import.meta.env.NG_APP_BACKEND}/api/strategy`;
  private portfolioApiUrl = `${import.meta.env.NG_APP_BACKEND}/api/portfolio`;

  constructor(private http: HttpClient) { }

  getSymbols(): Observable<{ symbols: string[] }> {
    return this.http.get<{ symbols: string[] }>(`${this.apiUrl}/symbols`);
  }

  getDropdowns(symbol: string): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/dropdowns/${symbol}`);
  }

  getSymbolData(symbol: string): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/symbol-data/${symbol}`);
  }

  getYearwiseData(symbol: string): Observable<YearwiseDataResponse> {
    return this.http.get<YearwiseDataResponse>(`${this.apiUrl}/yearwise-data/${symbol}`);
  }

  getFuturesData(symbol: string, expiry: string, identifier: string): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/futures-data/${symbol}?expiry=${expiry}&identifier=${identifier}`);
  }

  getOptionChain(symbol: string, expiry?: string): Observable<any> {
    let params = new HttpParams();
    if (expiry) {
      params = params.set('expiry', expiry);
    }
    return this.http.get<any>(`${this.apiUrl}/option-chain/${symbol}`, { params });
  }

  saveStrategy(strategy: Strategy): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/save`, strategy);
  }

  getUserStrategies(): Observable<UserStrategies> {
    return this.http.get<UserStrategies>(`${this.apiUrl}/user-strategies`);
  }

  deleteStrategy(id: number): Observable<any> {
    return this.http.delete<any>(`${this.apiUrl}/strategy/${id}`);
  }

  getPortfolioHoldings(): Observable<PortfolioSyncedResponse> {
    return this.http.get<PortfolioSyncedResponse>(`${this.portfolioApiUrl}/holdings`);
  }
}

File: mobile/lib/features/strategy_builder/strategy_service.dart
import 'dart:convert';
import '../../core/services/api_service.dart';
import '../../config/api_config.dart';

class StrategyPosition {
  final String segment;
  final String expiry;
  final double? strike;
  final String? optionType;
  final String action;
  final int qty;
  final double entryPrice;

  StrategyPosition({
    required this.segment,
    required this.expiry,
    this.strike,
    this.optionType,
    required this.action,
    required this.qty,
    required this.entryPrice,
  });

  StrategyPosition copyWith({
    String? segment,
    String? expiry,
    double? strike,
    String? optionType,
    String? action,
    int? qty,
    double? entryPrice,
  }) {
    return StrategyPosition(
      segment: segment ?? this.segment,
      expiry: expiry ?? this.expiry,
      strike: strike ?? this.strike,
      optionType: optionType ?? this.optionType,
      action: action ?? this.action,
      qty: qty ?? this.qty,
      entryPrice: entryPrice ?? this.entryPrice,
    );
  }

  Map<String, dynamic> toJson() => {
    'segment': segment,
    'expiry': expiry,
    'strike': strike,
    'option_type': optionType,
    'action': action,
    'qty': qty,
    'entry_price': entryPrice,
  };

  factory StrategyPosition.fromJson(Map<String, dynamic> json) => StrategyPosition(
    segment: json['segment'],
    expiry: json['expiry'],
    strike: json['strike']?.toDouble(),
    optionType: json['option_type'],
    action: json['action'],
    qty: json['qty'],
    entryPrice: json['entry_price'].toDouble(),
  );
}

class StrategyModel {
  final int? id;
  final String name;
  final String symbol;
  final List<StrategyPosition> positions;
  final String? createdAt;

  StrategyModel({
    this.id,
    required this.name,
    required this.symbol,
    required this.positions,
    this.createdAt,
  });

  Map<String, dynamic> toJson() => {
    'name': name,
    'symbol': symbol,
    'positions': positions.map((p) => p.toJson()).toList(),
  };

  factory StrategyModel.fromJson(Map<String, dynamic> json) => StrategyModel(
    id: json['id'],
    name: json['name'],
    symbol: json['symbol'],
    positions: (json['positions'] as List).map((p) => StrategyPosition.fromJson(p)).toList(),
    createdAt: json['created_at'],
  );
}

class StrategyService {
  final String _strategyUrl = '${ApiConfig.apiUrl}/strategy';

  Future<List<String>> getSymbols() async {
    final res = await ApiService.get('$_strategyUrl/symbols');
    if (res.statusCode == 200) {
      final data = json.decode(res.body);
      return List<String>.from(data['symbols']);
    }
    return [];
  }

  Future<Map<String, dynamic>> getDropdowns(String symbol) async {
    final res = await ApiService.get('$_strategyUrl/dropdowns/${symbol.toUpperCase()}');
    return json.decode(res.body);
  }

  Future<Map<String, dynamic>> getSymbolData(String symbol) async {
    final res = await ApiService.get('$_strategyUrl/symbol-data/${symbol.toUpperCase()}');
    return json.decode(res.body);
  }

  Future<Map<String, dynamic>> getFuturesData(String symbol, {String? expiry, String? identifier}) async {
    final params = <String, String>{};
    if (expiry != null && expiry.isNotEmpty) params['expiry'] = expiry;
    if (identifier != null && identifier.isNotEmpty) params['identifier'] = identifier;
    final res = await ApiService.get(
      '$_strategyUrl/futures-data/${symbol.toUpperCase()}',
      queryParams: params.isNotEmpty ? params : null,
    );
    return json.decode(res.body);
  }

  Future<Map<String, dynamic>> getOptionChain(String symbol, {String? expiry}) async {
    final res = await ApiService.get(
      '$_strategyUrl/option-chain/${symbol.toUpperCase()}',
      queryParams: expiry != null ? {'expiry': expiry} : null,
    );
    return json.decode(res.body);
  }

  Future<bool> saveStrategy(StrategyModel strategy) async {
    final res = await ApiService.post(
      '$_strategyUrl/save',
      body: strategy.toJson(),
    );
    return res.statusCode == 200;
  }

  Future<Map<String, List<StrategyModel>>> getUserStrategies() async {
    final res = await ApiService.get('$_strategyUrl/user-strategies');
    if (res.statusCode == 200) {
      final data = json.decode(res.body);
      return {
        'live': (data['live'] as List).map((s) => StrategyModel.fromJson(s)).toList(),
        'history': (data['history'] as List).map((s) => StrategyModel.fromJson(s)).toList(),
      };
    }
    return {'live': [], 'history': []};
  }

  Future<bool> deleteStrategy(int id) async {
    final res = await ApiService.delete('$_strategyUrl/$id');
    return res.statusCode == 200;
  }
}
